test/tuple.jl

```diff
@@ -783,3 +783,11 @@ namedtup = (;a=1, b=2, c=3)
 # Make sure that tuple iteration is foldable
 @test Core.Compiler.is_foldable(Base.infer_effects(iterate, Tuple{NTuple{4, Float64}, Int}))
 @test Core.Compiler.is_foldable(Base.infer_effects(eltype, Tuple{Tuple}))
+
+# some basic equivalence handling tests for Union{} appearing in Tuple Vararg parameters
+@test Tuple{} <: Tuple{Vararg{Union{}}}
+@test Tuple{Int} <: Tuple{Int, Vararg{Union{}}}
+@test_throws ErrorException("Tuple field type cannot be Union{}") Tuple{Int, Vararg{Union{},1}}
+@test_throws ErrorException("Tuple field type cannot be Union{}") Tuple{Vararg{Union{},1}}
+@test Tuple{} <: Tuple{Vararg{Union{},N}} where N
+@test !(Tuple{} >: Tuple{Vararg{Union{},N}} where N)
```

Comment:
Why is that? Is there an `N` for which the LHS is not `>:` the RHS? Isn't `N==0` the only admissible choice?

Reply:
Should we normalize `Tuple{A,B,C,Vararg{Union{},N}} where N` to `Tuple{A,B,C}`?

Reply:
No, that makes a different type, where N is free

Reply:
`Tuple{Vararg{Union{}}}` is invalid

Reply:
I guess that raises the possibility of making `Vararg{Union{}}` itself also directly throw an error. I was trying to be conservative in allowing Union{} but constructing a different sort of object. There could be a case to be made that Union{} should not appear their either though (which would allow deleting several more tests as non-applicable also)